src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -438,6 +488,8 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
             Location location,
             DiagnosticBag diagnostics)
         {
+            type = (NamedTypeSymbol)type.TupleUnderlyingTypeOrSelf();
```

Comment:
Re "TypeSymbolExtensions.VisitType" At first I thought that's the right way to do this but since all elements are flattened in long tuples I could not use VisitType to recursively check for constraints, instead I iterate over all elements and so on.

`// Individual elements are checked in tuple construction whence we have precise element locations.`

I'm not sure what's confusing here.. that seems ok to you?